src/main/java/org/candlepin/model/OwnerProductCurator.java

```diff
@@ -289,6 +290,52 @@ public CandlepinQuery<Product> getProductsByIdsUsingOwnerKey(String ownerKey,
         return this.cpQueryFactory.<Product>buildQuery();
     }
 
+    public Map<String, Set<String>> getSyspurposeByOwner(Owner owner) {
+        return this.getSyspurposeByOwner(owner.getId());
+    }
+
+    public Map<String, Set<String>> getSyspurposeByOwner(String ownerId) {
+        String sql =
+            "SELECT DISTINCT a.name, a.value  " +
+            "FROM cp2_product_attributes a JOIN cp_pool b ON a.product_uuid=b.product_uuid " +
+            "WHERE a.name IN ('usage','roles','addons','support_type') " +
+            "AND b.owner_id=:owner_id " +
+            "AND b.endDate > CURRENT_TIMESTAMP ";
+        List<Object[]> result = this.getEntityManager()
+            .createNativeQuery(sql)
+            .setParameter("owner_id", ownerId)
+            .getResultList();
+
+        sql =
+            "SELECT DISTINCT a.name, a.value " +
+            "FROM cp2_product_attributes a JOIN cp_pool b ON a.product_uuid=b.product_uuid " +
+            "WHERE name IN ('support_level') " +
+            "AND b.owner_id=:owner_id " +
+            "AND b.endDate > CURRENT_TIMESTAMP " +
+            "AND a.product_uuid NOT IN " +
+            "(SELECT a.product_uuid FROM cp2_product_attributes a " +
```

Comment:
Against stage it was not much faster and the highest number of products that any owner has is about 700.
I still made it a join instead of a subselect because I was already there.

Reply:
What do the query plans look like between the two?

Reply:
http://pastebin.test.redhat.com/1046481